{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deep mutational scanning of ZIKV E protein\n",
    "Deep mutational scanning of ZIKV E from the MR766 strain.\n",
    "Experiments performed by the [Matt Evans lab](http://labs.icahn.mssm.edu/evanslab/).\n",
    "Sequencing and computational analyses performed by the [Bloom lab](https://research.fhcrc.org/bloom/en.html).\n",
    "\n",
    "## Set up for analysis\n",
    "Import Python packages and modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using dms_tools2 2.3.0\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "import pandas\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "import dms_tools2\n",
    "from dms_tools2.ipython_utils import showPDF\n",
    "print(f\"Using dms_tools2 {dms_tools2.__version__}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify information about running analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_existing = 'yes' # use existing output\n",
    "\n",
    "ncpus = 4 # max CPUs to use\n",
    "\n",
    "# directories\n",
    "resultsdir = './results/'\n",
    "os.makedirs(resultsdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Input data found in the [./data/](data) directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseqfile = './data/E.fasta' # sequence of wildtype gene\n",
    "samplelist = './data/samplelist.csv' # samples sequenced\n",
    "alignspecsfile = './data/subamplicon_alignspecs.txt'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process deep sequencing data\n",
    "We process the data from the [barcoded subamplicon deep sequencing](https://jbloomlab.github.io/dms_tools2/bcsubamp.html) to count the frequency of each codon in each sample.\n",
    "\n",
    "First, we read in the samples:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>library</th>\n",
       "      <th>selection</th>\n",
       "      <th>antibody</th>\n",
       "      <th>percent_infectivity</th>\n",
       "      <th>R1</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>wildtype</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>wildtype-plasmid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>virus</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>virus</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>virus</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>wildtype</td>\n",
       "      <td>virus</td>\n",
       "      <td>none</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>wildtype-virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>no-antibody-A</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>71.06000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-no-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>no-antibody-B</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>131.89000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-no-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>no-antibody-C</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>97.05000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-no-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>no-antibody-A</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>132.77000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-no-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>no-antibody-B</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>82.35000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-no-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>no-antibody-C</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>84.87000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-no-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>no-antibody-A</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>86.42000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-no-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>no-antibody-B</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>98.49000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-no-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>no-antibody-C</td>\n",
       "      <td>no-antibody</td>\n",
       "      <td>115.10000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-no-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>control-antibody-A</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>104.65000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-control-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>control-antibody-B</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>119.27000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-control-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>control-antibody-C</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>108.40000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-control-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>control-antibody-A</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>67.35000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-control-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>control-antibody-B</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>82.77000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-control-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>control-antibody-C</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>112.47000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-control-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>control-antibody-A</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>80.14000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-control-antibody-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>control-antibody-B</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>86.85000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-control-antibody-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>control-antibody-C</td>\n",
       "      <td>control-antibody</td>\n",
       "      <td>93.19000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-control-antibody-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA64-A</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>2.34000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA64-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA64-B</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>3.01000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA64-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA64-C</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>2.35000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA64-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA64-A</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>1.54000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA64-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA64-B</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>1.62000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA64-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA64-C</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>1.76000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA64-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA64-A</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>2.21000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA64-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA64-B</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>1.67000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA64-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA64-C</td>\n",
       "      <td>ZKA64</td>\n",
       "      <td>1.48000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA64-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA185-A</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>0.00059</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA185-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA185-B</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA185-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib1</td>\n",
       "      <td>ZKA185-C</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib1-ZKA185-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA185-A</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>1.93000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA185-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA185-B</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>1.40000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA185-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib2</td>\n",
       "      <td>ZKA185-C</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>1.60000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib2-ZKA185-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA185-A</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>1.39000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA185-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA185-B</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>2.15000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA185-B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Lib3</td>\n",
       "      <td>ZKA185-C</td>\n",
       "      <td>ZKA185</td>\n",
       "      <td>2.11000</td>\n",
       "      <td>/fh/fast/bloom_j/SR/ngs/illumina/bloom_lab/180...</td>\n",
       "      <td>Lib3-ZKA185-C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "samples = (pandas.read_csv(samplelist)\n",
    "           .assign(name=lambda x: x.library + '-' + x.selection)\n",
    "           )\n",
    "\n",
    "display(HTML(samples.to_html(index=False)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we read in the alignment specs for the [barcoded subamplicon sequencing](https://jbloomlab.github.io/dms_tools2/bcsubamp.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1,303,33,38 304,609,38,40 610,903,41,36 904,1200,41,37 1201,1512,36,35\n"
     ]
    }
   ],
   "source": [
    "with open(alignspecsfile) as f:\n",
    "    alignspecs = f.read().strip()\n",
    "print(alignspecs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we use the [dms2_batch_bcsubamp](https://jbloomlab.github.io/dms_tools2/dms2_batch_bcsubamp.html) program to process the deep sequencing data to obtain codon counts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "countsdir = os.path.join(resultsdir, 'codoncounts')\n",
    "os.makedirs(countsdir, exist_ok=True)\n",
    "\n",
    "bcsubamp_batchfile = os.path.join(countsdir, 'batch.csv')\n",
    "samples[['name', 'R1']].to_csv(bcsubamp_batchfile, index=False)\n",
    "\n",
    "subprocess.check_call([\n",
    "    'dms2_batch_bcsubamp',\n",
    "    '--batchfile', bcsubamp_batchfile,\n",
    "    '--refseq', refseqfile,\n",
    "    '--alignspecs', alignspecs,\n",
    "    '--outdir', countsdir,\n",
    "    '--summaryprefix', 'summary',\n",
    "    '--R1trim', '200',\n",
    "    '--R2trim', '200',\n",
    "    '--ncpus', str(ncpus),\n",
    "    '--use_existing', use_existing\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
